src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java

```diff
@@ -40,26 +40,25 @@
     protected JiqsItemParser parser;
     protected List<Item> result = null;
     private int currentResultIndex = 0;
-    private boolean _isRDDInitialized = false;
-    private boolean _isRDD;
 
-    protected HybridRuntimeIterator(List<RuntimeIterator> children, IteratorMetadata iteratorMetadata) {
-        super(children, iteratorMetadata);
+    protected HybridRuntimeIterator(
+            List<RuntimeIterator> children,
+            ExecutionMode executionMode,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(children, executionMode, iteratorMetadata);
+        fallbackToRDDIfDFNotImplemented(executionMode);
         this.parser = new JiqsItemParser();
     }
 
-    @Override
-    public boolean isRDD() {
-        if (!_isRDDInitialized) {
-            _isRDD = initIsRDD();
-            _isRDDInitialized = true;
-        }
-        return _isRDD;
+    protected boolean implementsDataFrames() {
+        return false;
     }
 
-    @Override
-    public boolean isDataFrame() {
-        return false;
+    protected void fallbackToRDDIfDFNotImplemented(ExecutionMode executionMode) {
+        if (executionMode == ExecutionMode.DATAFRAME && !this.implementsDataFrames()) {
+            this._highestExecutionMode = ExecutionMode.RDD;
+        }
```

Comment:
I will investigate again why this function was needed in the first place, it may be masking some other correctness issue

Reply:
I think that's fine for now -- in the future, it would be desirable to have a perfect match between what the static analysis says and what actually happens at runtime, but we can fine-tune later.